pubsub/alicloud/rocketmq/rocketmq.go

```diff
@@ -0,0 +1,269 @@
+		r.logger.Errorf("rocketmq init consumer failed: %v", err)
+
+		return err
+	}
+
+	return nil
+}
+
+func (r *rocketMQ) setupPublisher() (mqw.Producer, error) {
+	if producer, ok := mqw.Producers[r.settings.AccessProto]; ok {
+		md := r.settings.ToRocketMQMetadata()
+		if err := producer.Init(md); err != nil {
+			r.logger.Debugf("rocketmq producer init failed: %v", err)
+
+			return nil, fmt.Errorf("setupPublisher failed. %w", err)
+		}
+		r.logger.Infof("rocketmq proto: %s", r.settings.AccessProto)
+
+		return producer, nil
+	}
+
+	return nil, errors.New("rocketmq error: cannot found rocketmq producer")
+}
+
+func (r *rocketMQ) setupConsumer() (mqw.PushConsumer, error) {
+	if consumer, ok := mqw.Consumers[r.settings.AccessProto]; ok {
+		md := r.settings.ToRocketMQMetadata()
+		if err := consumer.Init(md); err != nil {
+			r.logger.Errorf("rocketmq consumer init failed: %v", err)
+
+			return nil, fmt.Errorf("setupConsumer failed. %w", err)
+		}
+		r.logger.Infof("rocketmq access proto: %s", r.settings.AccessProto)
+
+		return consumer, nil
+	}
+
+	return nil, errors.New("rocketmq error: cannot found consumer")
+}
+
+func (r *rocketMQ) Publish(req *pubsub.PublishRequest) error {
+	msg := primitive.NewMessage(req.Topic, req.Data).WithTag(req.Metadata[metadataRocketmqTag]).
+		WithKeys([]string{req.Metadata[metadataRocketmqKey]})
+
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+	result, err := r.producer.SendSync(ctx, msg)
+
+	if result != nil {
+		r.logger.Debugf("rocketmq send result topic:%s tag:%s status:%v", req.Topic, msg.GetTags(), result.Status)
+	}
+
+	if err != nil {
+		r.logger.Errorf("error send message topic:%s : %v", req.Topic, err)
+
+		return fmt.Errorf("publish message failed. %w", err)
+	}
+
+	return nil
+}
+
+func (r *rocketMQ) addTopic(newTopic string, selector mqc.MessageSelector) []string {
+	// Add topic to our map of topics
+	r.topics[newTopic] = selector
+
+	topics := make([]string, len(r.topics))
+
+	i := 0
+	for topic := range r.topics {
+		topics[i] = topic
+		i++
+	}
+
+	return topics
+}
+
+// Close down consumer group resources, refresh once
+func (r *rocketMQ) closeSubscripionResources() {
+	if r.consumer != nil {
+		if len(r.topics) > 0 {
+			_ = r.consumer.Shutdown()
+		}
+	}
+}
+
+func (r *rocketMQ) Subscribe(req pubsub.SubscribeRequest, handler pubsub.Handler) error {
+	if req.Metadata == nil {
+		req.Metadata = make(map[string]string, 1)
+	}
+
+	consumerGroup := req.Metadata[metadataRocketmqConsumerGroup]
+	if len(consumerGroup) == 0 {
+		consumerGroup = r.settings.ConsumerGroup
+	}
+
+	mqType := req.Metadata[metadataRocketmqType]
+	mqExpr := req.Metadata[metadataRocketmqExpression]
+	if len(mqType) != 0 &&
+		(mqType != string(mqc.SQL92) &&
+			mqType != string(mqc.TAG)) {
+		r.logger.Warnf("rocketmq subscribe to topic %s failed because some illegal type(%s).", req.Topic, req.Metadata[metadataRocketmqType])
+
+		return nil
+	}
+
+	r.closeSubscripionResources()
```

Comment:
Is it required `r.consume` to be closed and re-setup each time Subscribe is called? I looks like its safe to just add new topics simply by calling `r.consumer.Subscribe`. Does subscribe not work after `r.consumer.Start` is called?

Reply:
yes，subscribe topics must called before  r.consumer.Start